File: 0279_Perfect_Squares.py
class Solution:
    def numSquares(self, n: 'int') -> 'int':
        dp = [0]
        sqrs = [x * x for x in range(1, int(math.sqrt(n)) + 1)]
        for i in range(1, n + 1):
            minimum = i
            for s in sqrs[:int(math.sqrt(i))]:
                minimum = min(minimum, dp[i - s] + 1)
            dp.append(minimum)
        return dp[-1]

File: 0279_Perfect_Squares.cpp
class Solution {
public:
    int numSquares(int n) {
        vector<int> dp (n + 1, INT_MAX);
        dp[0] = 0;
        dp[1] = 1;
        for (int i = 1; i <= n; ++i) {
            for (int s = 1; s * s <= i; ++s) {
                dp[i] = min(dp[i], dp[i - s * s] + 1);
            }
        }
        return dp[n];
    }
};
